### DISCUSSION TITLE
ABR GPU Acceleration.

### DISCUSSION THREAD

**Discussion by Pernifloss** [OvenMediaLabs/OvenMediaEngine#1113] [Help]
Hello !!

I'm trying to use OME ABR with GPU acceleration.

I'm on Windows 10 I have an Nvidia GeForce GTX 1050 Ti

I'm trying to run OME in Docker and make it use my GPU to encode/decode for ABR.

First, I built the docker image as specified in the documentation

`docker build --file Dockerfile -t airensoft/ovenmediaengine:dev --build-arg GPU=TRU`
Build was successful and I didn't notice any errors.

I run it using `--gpus all`
 
Then on docker logs I have this line : 
`[2023-02-08 14:42:52.451] W [OvenMediaEngine:1] Transcoder | transcoder_application.cpp:45   | There is no supported hardware accelerator`

When I stream, ABR works, I have my 3 stream quality in OvenMedia Player: 
 
![image](https://user-images.githubusercontent.com/22869959/217572942-dbf793a2-8cea-4a52-9ed1-f9b63f55eee2.png)

But it is encoding on CPU.

---

Here is my output profile

                       <OutputProfiles>
                        <HardwareAcceleration>true</HardwareAcceleration>
                        <OutputProfile>
                            <Name>bypass_stream</Name>
                            <OutputStreamName>${OriginStreamName}</OutputStreamName>
                            <Playlist>
                                <Name>For LLHLS</Name>
                                <FileName>abr</FileName>
                                <Options> 
                                    <WebRtcAutoAbr>true</WebRtcAutoAbr>
                                </Options>
                                <Rendition>
                                    <Name>Bypass</Name>
                                    <Video>bypass_video</Video>
                                </Rendition>
                                <Rendition>
                                    <Name>SD</Name>
                                    <Video>video_540</Video>
                                </Rendition>
                                <Rendition>
                                    <Name>HD</Name>
                                    <Video>video_720</Video>
                                </Rendition>
                            </Playlist>
                            <Playlist>
                                <Name>Change Default</Name>
                                <FileName>llhls</FileName>
                                <Rendition>
                                    <Name>HD</Name>
                                    <Video>video_720</Video>
                                </Rendition>
                            </Playlist>
                            <Encodes>
                                <Video>
                                    <Name>bypass_video</Name>
                                    <Bypass>true</Bypass>
                                </Video>
                                <Video>
                                    <Name>video_540</Name>
                                    <Codec>h264</Codec>
                                    <Bitrate>1024000</Bitrate>
                                    <Framerate>30</Framerate>
                                    <Width>1024</Width>
                                    <Height>576</Height>
                                    <Preset>faster</Preset>
                                </Video>
                                <Video>
                                    <Name>video_720</Name>
                                    <Codec>h264</Codec>
                                    <Bitrate>2024000</Bitrate>
                                    <Framerate>30</Framerate>
                                    <Width>1280</Width>
                                    <Height>720</Height>
                                    <Preset>faster</Preset>
                                </Video>
                            </Encodes>
                        </OutputProfile>
                    </OutputProfiles>

---

I think I managed to install NVIDIA Container Toolkit.

When I run 
` docker run --gpus all nvidia/cuda:11.8.0-base-ubuntu22.04 nvidia-smi`

I get 

```

    Wed Feb  8 15:18:25 2023                                                       
    +-----------------------------------------------------------------------------+
    | NVIDIA-SMI 525.65       Driver Version: 527.56       CUDA Version: 12.0     |
    |-------------------------------+----------------------+----------------------+
    | GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
    | Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
    |                               |                      |               MIG M. |
    |===============================+======================+======================|
    |   0  NVIDIA GeForce ...  On   | 00000000:01:00.0  On |                  N/A |
    |  0%   41C    P8    N/A /  75W |   2046MiB /  4096MiB |     31%      Default |
    |                               |                      |                  N/A |
    +-------------------------------+----------------------+----------------------+
                                                                                   
    +-----------------------------------------------------------------------------+
    | Processes:                                                                  |
    |  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
    |        ID   ID                                                   Usage      |
    |=============================================================================|
    |    0   N/A  N/A        23      G   /Xwayland                       N/A      |
    |    0   N/A  N/A        34      G   /Xwayland                       N/A      |
    |    0   N/A  N/A        37      G   /Xwayland                       N/A      |
    +-----------------------------------------------------------------------------+

```

---

I'm streaming with OBS encoding with Nvidia Nvenc H.264.
Obs is using my GPU correctly.

---

Did I miss a step or something? 
Thanks for your time !

**Comment by getroot** [comment]
One of your build parameters, GPU=TRU, seems to be a typo in the documentation. Could you try again with GPU=TRUE?

**Comment by Pernifloss** [comment]
Hello again, with `GPU=TRUE` docker build fail with this error : 

```
#8 470.1 ERROR: failed checking for nvcc.
#8 470.1
#8 470.1 If you think configure made a mistake, make sure you are using the latest
#8 470.1 version from Git.  If the latest version fails, report the problem to the
#8 470.1 ffmpeg-user@ffmpeg.org mailing list or IRC #ffmpeg on irc.libera.chat.
#8 470.1 Include the log file "ffbuild/config.log" produced by configure as this will help
#8 470.1 solve the problem.
#8 470.1 (ffmpeg) installation has failed.
------
executor failed running [/bin/sh -c if [ "$GPU" = "TRUE" ] ; then                 ${TEMP_DIR}/misc/install_nvidia_docker_image.sh ;                 ${TEMP_DIR}/misc/prerequisites.sh  --enable-nvc ;         else                 ${TEM
P_DIR}/misc/prerequisites.sh ;         fi]: exit code: 1

```

**Comment by Keukhan** [comment]
@Pernifloss 
The last docker image had a problem with the nvcc compiler missing. so, Dockerfile has been updated to include nvcc (CUDA Compiler).
Try the test again. I will wait for your results. 

Thanks.

**Comment by Pernifloss** [comment]
Hello @Keukhan Thanks for your response !!

I pulled last git repo version (master branch), the built went well, I didn't notice any error,

but when I try to run my image, I have this error message:

```
 docker run -d -p 1935:1935 -p 3333:3333 --gpus all airensoft/ovenmediaengine:dev
5602f3e90021f45fc7efdd3af6c6e009ab6060eeb3eceebe856a822e1aabdd32
docker: Error response from daemon: failed to create shim task: OCI runtime create failed: runc create failed: unable to start container process: error during container init: error running hook #0: error running hook: exit status 1, stdout: , stderr: Auto-detected mode as 'legacy'
nvidia-container-cli: mount error: file creation failed: /var/lib/docker/overlay2/34e8e4de89a6354421c1eeabf3c4ef648fe2578fcc734639b73ea862e376e8e7/merged/usr/lib/x86_64-linux-gnu/libnvidia-ml.so.1: file exists: unknown.
```

Thanks.
